# Data Consistency Checks

1. Import Libraries and Data

2. Consistency Checks on Dataframes

3. Export Data

In [2]:
import pandas as pd
import numpy as np
import os

In [6]:
df_prods = pd.read_csv(r'C:\Users\rpa21\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [8]:
path = r'C:\Users\rpa21\Instacart Basket Analysis'

In [9]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [10]:
df_ords = pd.read_csv(r'C:\Users\rpa21\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [11]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

02. Creating New Dataframe 

In [12]:
#Create a dataframe
df_test = pd.DataFrame()

In [13]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [14]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [16]:
# Check for mix types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [17]:
df_test['mix'] = df_test['mix'].astype('str')

3. Finding Missing Values

In [18]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [19]:
# A new dataframe with only values in the product_name column wiht missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [ ]:
df_nan

In [21]:
# Look at the shape of the dataframe. 
df_prods.shape

(49693, 5)

In [22]:
# Creating a new clean dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
# Checking if clean
df_prods_clean.shape

(49677, 5)

In [24]:
# Finding full duplicates in the dataframe.
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
# Looking at the duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [26]:
# Dropping duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
# Checking if duplicates are removed
df_prods_clean_no_dups.shape

(49672, 5)

In [28]:
# Export Dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))



Task 4.5 on Ords Dataframe

In [ ]:
#2. Run the df.describe() function on your df_ords.  Does anything about the data look
#off or should be investigated further.

df_ords.describe()

The Max value of the days_since_last_order looks lower than one would expect.

In [31]:
# 3. Check for mixed-type data in your df_ords dataframe. 

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

The columns do not have mixed data

In [32]:
# 5. Run a check for missing values in your df_ords dataframe.

df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

There are 206,209 values missing in the days_since_last_order column. 
This could be the result of data corruption or the values were never entered.

In [33]:
# 6. Address the missing values using an appropriate method.

With many observations missing, I feel leaving this data in the dataframe would be the best 
course of action. One possiblity is to create a new column but it is such a large amount 
of the data, it may affect analysis. 

In [34]:
# 7. Run a check for duplicate values in your df_ords data.

df_dups = df_ords[df_ords.duplicated()]

In [36]:
# Looking at duplicates

df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

8. There are no duplicate values that have been found. 

In [39]:
# 9. Export to Prepared Data

In [40]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))